stacks-signer/src/signerdb.rs

```diff
@@ -961,35 +827,63 @@ impl SignerDb {
     }
 
     /// Either instantiate a new database, or migrate an existing one
-    /// If the detected version of the existing database is 0 (i.e., a pre-migration
-    /// logic DB, the DB will be dropped).
     fn create_or_migrate(&mut self) -> Result<(), DBError> {
         self.register_scalar_functions()?;
         let sql_tx = tx_begin_immediate(&mut self.db)?;
-        loop {
-            let version = Self::get_schema_version(&sql_tx)?;
-            match version {
-                0 => Self::schema_1_migration(&sql_tx)?,
-                1 => Self::schema_2_migration(&sql_tx)?,
-                2 => Self::schema_3_migration(&sql_tx)?,
-                3 => Self::schema_4_migration(&sql_tx)?,
-                4 => Self::schema_5_migration(&sql_tx)?,
-                5 => Self::schema_6_migration(&sql_tx)?,
-                6 => Self::schema_7_migration(&sql_tx)?,
-                7 => Self::schema_8_migration(&sql_tx)?,
-                8 => Self::schema_9_migration(&sql_tx)?,
-                9 => Self::schema_10_migration(&sql_tx)?,
-                10 => Self::schema_11_migration(&sql_tx)?,
-                11 => Self::schema_12_migration(&sql_tx)?,
-                12 => Self::schema_13_migration(&sql_tx)?,
-                13 => Self::schema_14_migration(&sql_tx)?,
-                14 => break,
-                x => return Err(DBError::Other(format!(
-                    "Database schema is newer than supported by this binary. Expected version = {}, Database version = {x}",
-                    Self::SCHEMA_VERSION,
-                ))),
+
+        let mut current_db_version = Self::get_schema_version(&sql_tx)?;
+        debug!("Current SignerDB schema version: {}", current_db_version);
+
+        for migration in MIGRATIONS.iter() {
+            if current_db_version < migration.version {
+                if current_db_version != migration.version - 1 {
+                    // This implies a gap or out-of-order migration definition,
+                    // or the database is at a version X, and the next migration is X+2 instead of X+1.
+                    sql_tx.rollback()?;
```

Comment:
Inverting the check on that first if statement I think makes things a little more clear:
```suggestion
        for migration in MIGRATIONS.iter() {
            if current_db_version >= migration.version {
                // don't need this migration, continue to see if we need later migrations
                continue;
            }
            if current_db_version != migration.version - 1 {
                // This implies a gap or out-of-order migration definition,
                // or the database is at a version X, and the next migration is X+2 instead of X+1.
                sql_tx.rollback()?;
```